### PR TITLE
Stop using Julia's size classes when using MMTk

### PR DESCRIPTION
**Pull request by udesou**
When allocating an object, we are still allocating the number of bytes according to Julia's size classes, which may be an overestimation of the number of bytes an object actually requires. This PR removes this since this is specific to Julia's stock GC.

### CODE REVIEW

File: src/mmtk-gc.c [hunk 1]
```diff
@@ -53,6 +53,17 @@ static inline void malloc_maybe_collect(jl_ptls_t ptls, size_t sz)
     }
 }
 
+// allocation
+int jl_gc_classify_pools(size_t sz, int *osize)
```
Comment (qinsoon) [inline src/mmtk-gc.c:57]: This method is still used in code gen for allocation fastpath. So allocation from generated allocation fastpath still uses pool size. What is your plan on that?

If we no longer use pool size in code gen, this method can be removed from `gc.h`.

Comment (qinsoon) [reply src/mmtk-gc.c:57]: My bad. It seems you changed the implementation of this method for MMTk. It calculates the `osize` differently, but still returns a pool offset. The pool offset will decide whether the compiler generates a call for big alloc or generates the sequence for fastpath alloc.

File: src/mmtk-gc.c [hunk 2]
```diff
@@ -53,6 +53,17 @@ static inline void malloc_maybe_collect(jl_ptls_t ptls, size_t sz)
...
+{
+    if (sz > GC_MAX_SZCLASS)
+        return -1;
+    size_t allocsz = sz + sizeof(jl_taggedvalue_t);
+    int klass = jl_gc_szclass(allocsz);
+    *osize = LLT_ALIGN(allocsz, 16);
+    return (int)(intptr_t)(&((jl_ptls_t)0)->heap.norm_pools[klass]);
```
Comment (qinsoon) [inline src/mmtk-gc.c]: Consider either removing `jl_gc_classify_pools` (and modifying the code in `llvm-final-gc-lowering.cpp`), or modifying the code in this method so it does not rely on the Julia size class. Maybe you also want to change the check `if (sz > GC_MAX_SZCLASS)` to use a large object threshold from MMTk (I don't know if Julia would work fine with it).

```suggestion
    return 0; // return a non-negative value, to indicate
```

File: src/mmtk-gc.c (hunk 1 above)
Comment (udesou) [reply src/mmtk-gc.c:57]: I changed that function a bit and added a few comments, but indeed, I'm still using it to decide whether to do a big alloc or a fastpath/slowpath one. But I'm not using the size classes in there anymore.

Comment (qinsoon) [reply src/mmtk-gc.c:57]: Yeah. The code looks good. I am not sure if you want to change the large object threshold from `GC_MAX_SZCLASS` to a value from MMTk. I am okay with the current code.

**Comment by qinsoon** [comment]
This PR aligns everything to 16 bytes. As the smallest size class is 8 bytes, we may allocate more memory than before, and could cause slowdown. MMTk assumes the allocation size is a multiple of `VM::MIN_ALIGNMENT`, which is defined as `4` for Julia (https://github.com/mmtk/mmtk-julia/blob/baeffd2d4e809289ecd962795142dd02bf59f8a7/mmtk/src/lib.rs#L43). Aligning to 16 bytes may be unnecessary.

**Comment by udesou** [comment]
> This PR aligns any object size to 16 bytes. As the smallest size class is 8 bytes, we may allocate more memory than before, and could cause slowdown. MMTk assumes the allocation size is a multiple of `VM::MIN_ALIGNMENT`, which is defined as `4` for Julia (https://github.com/mmtk/mmtk-julia/blob/baeffd2d4e809289ecd962795142dd02bf59f8a7/mmtk/src/lib.rs#L43). Aligning to 4 bytes should be enough for MMTk, and aligning to 16 bytes may be unnecessary unless Julia requires it.

8-byte aligned classes are used for strings: https://github.com/mmtk/julia/blob/5c406d9bb20d76e2298a6101f171cfac491f651c/src/julia_internal.h#L371. I think everything else is aligned to 16. I'll change the code to make sure I use that alignment only for strings.

**Comment by qinsoon** [comment]
> > This PR aligns any object size to 16 bytes. As the smallest size class is 8 bytes, we may allocate more memory than before, and could cause slowdown. MMTk assumes the allocation size is a multiple of `VM::MIN_ALIGNMENT`, which is defined as `4` for Julia (https://github.com/mmtk/mmtk-julia/blob/baeffd2d4e809289ecd962795142dd02bf59f8a7/mmtk/src/lib.rs#L43). Aligning to 4 bytes should be enough for MMTk, and aligning to 16 bytes may be unnecessary unless Julia requires it.
> 
> 8-byte aligned classes are used for strings:
> 
> https://github.com/mmtk/julia/blob/5c406d9bb20d76e2298a6101f171cfac491f651c/src/julia_internal.h#L371
> 
> I think everything else is aligned to 16. I'll change the code to make sure I use that alignment only for strings.

I am actually a bit confused by the comments in the code above ("the 8-byte aligned pools are only used for Strings"). They have 8 bytes aligned pools all the way up to 136 bytes. Does that mean all the objects in that size range are strings? That doesn't sound right.

**Comment by qinsoon** [comment]
For the current code, 
* String size is aligned up to 16 bytes (but this will be fixed so string size is unchanged -- aligned to 8 bytes).
* Objects between 16 bytes to 136 bytes should have the same allocation size as before.
* Objects larger than 144 bytes may have a smaller size.

**Review by qinsoon — APPROVED** [review]
LGTM

**Comment by qinsoon** [comment]
Can you get performance results for the PR (comparing with `master`)?

**Comment by udesou** [comment]
> Can you get performance results for the PR (comparing with `master`)?

Yes, I was doing it already, without considering the latest changes, but I'll re-run everything and put the results here once I get them.

**Pull request merged**